include/lsst/partition/ObjectIndex.h

```diff
@@ -0,0 +1,195 @@
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+/// \file
+/// \brief Operations with the "secondary" index
+
+#ifndef LSST_PARTITION_OBJECTINDEX_H
+#define LSST_PARTITION_OBJECTINDEX_H
+
+// System headers
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+
+// LSST headers
+#include "lsst/partition/Csv.h"
+
+// Forward declarations
+namespace lsst {
+namespace partition {
+    class ChunkLocation;
+}} // namespace lsst::partition
+
+// This header declarations
+namespace lsst {
+namespace partition {
+
+/**
+ * Class ObjectIndex is a singleton serving as a front-end to a specific implementation of
+ * the "secondary" index.
+ * 
+ * Objects of the class can be open in two modes:
+ * - WRITE: for writing into an index file at the specified location.
+ * - READ: reading from an index via the specified source.
+ * 
+ * Files written in the WRITE mode are open in the "append" mode, hence their previoous
+ * content won't be truncated. It's up to a client to ensure the files get removed if
+ * the clean starting state is required.
+ * 
+ * The source specification in the READ node is a URL-style resource, which has the following
+ * general syntax:
+ * @code
+ *   <technology>:<technolog-specific-details>
+ * @code
+ * Presently, only the file-based technology is supported. Hence, the file URL is expected
+ * to look like this:
+ * @code
+ *   file:<path-to-the-index-file>
+ * @code
+ * Normally the path specified in the resource would be a file which was produced
+ * in the WRITE mode. The path could have a relative or absolute value. It could also be
+ * located on a local or remote filesystem. Since the file would be read sequentially then
+ * no specific requirements for file locking or direct access are imposed by this implemenation.
+ *
+ * Notes on the thread-safety of the class's implementation:
+ * 1) All public methods of the class are thread safe.
+ * 2) An internal state of the index is guarded by a mutex.
+ */
+class ObjectIndex {
+public:
+    // Objects of this class can't be copied or directly constructed by clients.
+    ObjectIndex(ObjectIndex const&) = delete;
+    ObjectIndex& operator=(ObjectIndex const&) = delete;
+
+    /// The non-trivial destructor is needed to close and release resources used by the index.
+    ~ObjectIndex();
+
+    /// Modes for opening the index.
+    enum Mode {
+        READ,
+        WRITE
+    };
+
+    /**
+     * The factory method.
+     * 
+     * @return  A reference to the underlining implementation of the index
+     */
+    static ObjectIndex& instance() {
+        static ObjectIndex oi;
+        return oi;
+    }
+
+    /// @return A value of 'true' if the index is open or created.
+    bool isOpen() const { return _isOpen; }
```

Comment:
For unit testing

Reply:
If you do not create instances in non-open state then you do not even need to unit-test it 🙂

Reply:
Actually, when I started working on this implementation I tried to analyze where would be the best location for creating & opening the index. I could potentially make it in `main()`. In this case the index object could be fully initialized before starting that map-reduce machinery. Unfortunately that would break encapsulation of knowledge about parameters needed by class `Worker`, and I would end up with code spread in two locations. That would further complicate maintaining this application. Hence, I ended up with the "lazy" initialization logic made by the very first client Worker (thread).
Unfortunately, I can't do the same for the `ObjectIndex::close()` operation which should only be done after all worker threads have finished. That's why closing the index (flushing buffers into the index file, releasing resources) is still die in `main` as an explicit step.

Reply:
If closing happens naturally in `main()` than opening that resource should also be naturally done in `main()`. It's easier to track resource lifetime from a single unit rather than spreading it across multiple places.

Reply:
There are two decisions to be made by the application:
* which of two partitioning modes to employ: `RA/DEC` or `objectId`-based when processing the input files
* and, in what to do about the index: "create" the one or "open" the one.
Right now both decisions are done by the constructor of class `Worker` based on the corresponding values of the the parameters' group `part.*` (`part.pos`, `part.id`, and `part.id-url`). And not only that, one may also notice that injecting the corresponding configuration parameters into the Boost parser is done by a dedicated static method of class `Worker`. This kind of encapsulation is perfectly consistent with the design of the "map-reduce" machinery which assumes that worker classes (template parameters of class Job`) are self-sufficient. If we break that then we would have to spread the logic between class `Worker` and `main`, thus breaking the design of "map-reduce".